web/src/ui/timeline/content/useMediaContent.tsx

```diff
@@ -23,18 +24,65 @@ import DownloadIcon from "@/icons/download.svg?react"
 export const useMediaContent = (
 	content: MediaMessageEventContent, evtType: EventType, containerSize?: ImageContainerSize,
 ): [React.ReactElement | null, string, CSSProperties] => {
+	const imgEl = React.useRef<HTMLImageElement>(null)
+	const [loaded, setLoaded] = React.useState(false)
+	const onImageLoaded = () => setLoaded(true)
+
+	const blurhashEl = React.useRef<Blurhash>(null)
+	const [spoilerShowing, setSpoilerShowing] = React.useState(false)
+
+	React.useEffect(() => {
+		const imgElCurrent = imgEl.current
+
+		if (imgElCurrent) {
+			imgElCurrent.addEventListener("load", onImageLoaded)
+			return () => imgElCurrent.removeEventListener("load", onImageLoaded)
+		}
+	}, [imgEl])
+
 	const mediaURL = content.file?.url ? getEncryptedMediaURL(content.file.url) : getMediaURL(content.url)
 	const thumbnailURL = content.info?.thumbnail_file?.url
 		? getEncryptedMediaURL(content.info.thumbnail_file.url) : getMediaURL(content.info?.thumbnail_url)
 	if (content.msgtype === "m.image" || evtType === "m.sticker") {
 		const style = calculateMediaSize(content.info?.w, content.info?.h, containerSize)
-		return [<img
-			loading="lazy"
-			style={style.media}
-			src={mediaURL}
-			alt={content.filename ?? content.body}
-			onClick={use(LightboxContext)}
-		/>, "image-container", style.container]
+
+		const blurhash = content.info ? content.info["xyz.amorgan.blurhash"] : undefined
+		const isSpoiler = content["m.content_warning"]?.type === "m.spoiler" ||
+			content["town.robin.msc3725.content_warning"]?.type === "town.robin.msc3725.spoiler"
+		const showPlaceholder = (blurhash && !loaded) || (isSpoiler && !spoilerShowing)
```

Comment:
Not sure if this is the correct place, seems like it should be applied to all media instead of just images